compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/JvmInlineClassLowering.kt

```diff
@@ -0,0 +1,402 @@
+        listOfNotNull(
+            function.dispatchReceiverParameter?.let { it to expression.dispatchReceiver },
+            function.extensionReceiverParameter?.let { it to expression.extensionReceiver }
+        ) + function.valueParameters.map { it to expression.getValueArgument(it.index) }
+
+    private fun IrMemberAccessExpression.buildReplacement(
+        originalFunction: IrFunction,
+        original: IrMemberAccessExpression,
+        replacement: IrReplacementFunction
+    ) {
+        copyTypeArgumentsFrom(original)
+        for ((parameter, argument) in typedArgumentList(originalFunction, original)) {
+            if (argument == null) continue
+            val newParameter = replacement.valueParameterMap[parameter.symbol] ?: continue
+            putArgument(
+                replacement.function,
+                newParameter,
+                argument.transform(this@JvmInlineClassLowering, null)
+            )
+        }
+    }
+
+    private fun buildReplacementCall(
+        originalFunction: IrFunction,
+        original: IrFunctionAccessExpression,
+        replacement: IrReplacementFunction
+    ) = context.createIrBuilder(original.symbol)
+        .irCall(replacement.function)
+        .apply { buildReplacement(originalFunction, original, replacement) }
+
+    override fun visitFunctionReference(expression: IrFunctionReference): IrExpression {
+        val replacement = manager.getReplacementFunction(expression.symbol.owner)
+            ?: return super.visitFunctionReference(expression)
+        val function = replacement.function
+
+        return IrFunctionReferenceImpl(
+            expression.startOffset, expression.endOffset, expression.type,
+            function.symbol, function.descriptor,
+            function.typeParameters.size, function.valueParameters.size,
+            expression.origin
+        ).apply {
+            buildReplacement(expression.symbol.owner, expression, replacement)
+        }
+    }
+
+    override fun visitFunctionAccess(expression: IrFunctionAccessExpression): IrExpression {
+        val function = expression.symbol.owner
+        val replacement = manager.getReplacementFunction(function)
+            ?: return super.visitFunctionAccess(expression)
+        return buildReplacementCall(function, expression, replacement)
+    }
+
+    private fun coerceInlineClasses(argument: IrExpression, from: IrType, to: IrType) =
+        IrCallImpl(UNDEFINED_OFFSET, UNDEFINED_OFFSET, to, context.ir.symbols.unsafeCoerceIntrinsicSymbol).apply {
+            putTypeArgument(0, from)
+            putTypeArgument(1, to)
+            putValueArgument(0, argument)
+        }
+
+    private fun visitPrimaryInlineClassConstructorCall(expression: IrMemberAccessExpression, constructor: IrConstructor): IrExpression {
+        val parameter = constructor.valueParameters.single()
+        val arg = expression.getValueArgument(0) ?: parameter.defaultValue!!.expression.deepCopyWithSymbols()
```

Comment:
Loading parameter's default value won't work with an inline class loaded from dependencies. Can't we simply leave this call without an argument if it didn't have one, since `defaultArgumentStubPhase` is running right after this lowering?

Reply:
That's a good point. The problem here is that calls to the inline class constructor are replaced by the `unsafeCoerce` intrinsic, which is not treated in `defaultArgumentStubPhase`.

The way this is handled in the current backend is by always creating a `constructor-impl` function for the primary constructor (which doesn't do anything except nullability checks). This function is in turn lowered by `defaultArgumentStubPhase` which produces the correct behavior.

It's probably a good idea to reproduce this behavior anyway, because otherwise we might miss some null checks. I've added this to the IR lowering.

Reply:
Great, thanks. Note that in the current implementation the bytecode produced for `constructor-impl` after this lowering contains an unnecessary creation of the inline class:
```
    Code:
      stack=3, locals=1, args_size=1
         0: aload_0
         1: ldc           #71                 // String s
         3: invokestatic  #17                 // Method kotlin/jvm/internal/Intrinsics.checkParameterIsNotNull:(Ljava/lang/Object;Ljava/lang/String;)V
         6: new           #2                  // class Z
         9: dup
        10: aload_0
        11: invokespecial #73                 // Method "<init>":(Ljava/lang/String;)V
        14: invokevirtual #38                 // Method "unbox-impl":()Ljava/lang/String;
        17: areturn
```

But I suppose it can be fixed later

Reply:
Thank you for looking at the generated code. I've just fixed it and will add a test.

Reply:
Done